c_common/models/live_packet_gather/src/live_packet_gather.c

```diff
@@ -113,52 +114,54 @@ static struct lpg_config config;
-static inline void write_unaligned(
-        void *base, uint32_t index, uint32_t value) {
+static inline void write_word(void *base, uint32_t index, uint32_t value) {
     uint16_t *ary = base;
     uint32_t idx = index * 2;
     ary[idx++] = CLAMP16(value);
     ary[idx] = CLAMP16(value >> 16);
 }
 
+//! \brief Simple mirror of write_word() for true 16 bit values.
+static inline void write_short(void *base, uint32_t index, uint32_t value) {
+    uint16_t *ary = base;
+    ary[index] = CLAMP16(value);
+}
+
+//! \brief Get how many events there are waiting to be sent
+static inline uint8_t get_event_count(void) {
+    uint8_t event_count = buffer_index;
+    // If there are payloads, it takes two buffer values to encode them
+    if (HAVE_PAYLOAD(config.packet_type)) {
+        event_count >>= 1;
+    }
+    return event_count;
+}
+
 static void flush_events(void) {
     // Send the event message only if there is data
-    if (buffer_index > 0) {
-        uint8_t event_count;
-
-        if ((config.packets_per_timestamp == 0)
-                || (packets_sent < config.packets_per_timestamp)) {
-            // Get the event count depending on if there is a payload or not
-            if (HAVE_PAYLOAD(config.packet_type)) {
-                event_count = buffer_index >> 1;
+    if ((buffer_index > 0) && (
+            (config.packets_per_timestamp == 0) ||
+            (packets_sent < config.packets_per_timestamp))) {
+        // Get the event count depending on if there is a payload or not
+        uint8_t event_count = get_event_count();
+
+        // insert appropriate header
+        sdp_msg_aer_header[0] = temp_header | CLAMP8(event_count);
+
+        g_event_message.length =
+                sizeof(sdp_hdr_t) + header_len + event_count * event_size;
+
+        // Add the timestamp if required
+        if (sdp_msg_aer_payload_prefix && config.payload_timestamp) {
+            if (!HAVE_WIDE_LOAD(config.packet_type)) {
+                write_short(sdp_msg_aer_payload_prefix, 0, time);
             } else {
-                event_count = buffer_index;
+                write_word(sdp_msg_aer_payload_prefix, 0, time);
             }
-
-            // insert appropriate header
-            sdp_msg_aer_header[0] = temp_header | CLAMP8(event_count);
-
-            g_event_message.length =
-                    sizeof(sdp_hdr_t) + header_len + event_count * event_size;
-
-            if (config.payload_apply_prefix && config.payload_timestamp) {
-                if (!HAVE_WIDE_LOAD(config.packet_type)) {
-                    uint16_t *temp = sdp_msg_aer_payload_prefix;
-                    temp[0] = CLAMP16(time);
-                } else {
-                    write_unaligned(sdp_msg_aer_payload_prefix, 0, time);
-                }
-            }
-
-            spin1_send_sdp_msg(&g_event_message, 1);
-            packets_sent++;
         }
 
-        // reset packet content
-        uint16_t words_to_clear = (buffer_index * event_size) >> 2;
```

Comment:
This line was the actual place of the bug; when there are payloads, `buffer_index` is twice as large as when there are no payloads, so `words_to_clear` ends up twice as large as it should be. With enough packets, we end up overwriting memory outside the buffer…

Reply:
having looked at it (noted its disappeared in the new version) im not sure how the heck that ever worked for anything OTHER than things with event_size = 4. if i understand correctly. buffer_index is basically how many uint16's are in the circular buffer (which is likely the same as buffer->real_size / 4) so unless the mul and the shift cancel each other out. i can def see how that will misalign quickly. 

I'm going to presume that we got lucky for so long.